jax/experimental/mesh_utils.py

```diff
@@ -245,13 +246,24 @@ def _create_device_mesh(process_0_devices, global_devices, device_kind,
                         mesh_shape: Sequence[int],
                         contiguous_submeshes: bool = False) -> np.ndarray:
   # TODO(zhangqiaorjc): Handle TPU versions other than v4 more generally.
-  if device_kind == _TPU_V3:
-    device_mesh = np.asarray(global_devices).reshape(mesh_shape)
-    if mesh_shape[-1] == 8:
-      logging.info('Re-order TPUv3 device mesh for better performance.')
+  if device_kind in (_TPU_V2, _TPU_V3):
+    if len(global_devices) == 8:
+      logging.info('Reordering mesh to physical ring order on single-tray TPU v2/v3.')
+      device_mesh = np.asarray(global_devices)
+      device_mesh = device_mesh[np.array([0, 1, 2, 3, 6, 7, 4, 5])]
```

Comment:
might be useful to assign a variable for this ring order transpose array since it's used at two places and easy to get wrong accidentally